### PR TITLE
KVStore: Refine parallel prehandle snapshot (part-1)

### PR DESCRIPTION
**Pull request by JaySon-Huang**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflash/issues/8081

Problem Summary:

I'd like to introduce a new layer `SnapshotSSTReader` and move `getApproxBytes`/`findSplitKeys` into that layer. This could make the logic in parallel prehandle snapshots more clear.

This PR is a preceding PR that refactors the code to simplify the following logical change.

New `SnapshotSSTReader` interface in later PR:
```
struct SnapshotSSTReader
{
    SnapshotSSTReader(
        const SSTViewVec & snaps,
        const TiFlashRaftProxyHelper * proxy_helper,
        RegionID region_id,
        UInt64 snapshot_index,
        const ImutRegionRangePtr & region_range,
        std::optional<SSTScanSoftLimit> && soft_limit,
        const String & log_prefix);

    // Currently it only takes effect if using tablet sst reader which is usually a raftstore v2 case.
    // Otherwise will return zero.
    size_t getApproxBytes() const;
    std::vector<std::string> findSplitKeys(size_t splits_count) const;

    using SSTReaderPtr = std::unique_ptr<SSTReader>;
    SSTReaderPtr write_cf_reader;
    SSTReaderPtr default_cf_reader;
    SSTReaderPtr lock_cf_reader;
};
```

### What is changed and how it works?

```commit-message

```

* Unify `extra_id` by `split_id`
* Unify the return type of `executeParallelTransform` and `executeTransform` to be `std::tuple<ReadFromStreamResult, PrehandleResult>`
* Apply "early return" on `executeParallelTransform` to simplify the code indent layer

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp [hunk 1]
```diff
@@ -522,107 +523,107 @@ void executeParallelTransform(
...
+    const DM::SSTScanSoftLimit head_soft_limit(
+        DM::SSTScanSoftLimit::HEAD_OR_ONLY_SPLIT,
+        std::string(""),
+        std::string(split_keys[0]));
+    sst_stream->resetSoftLimit(head_soft_limit.clone());
 
     ParallelPrehandleCtxPtr parallel_ctx = std::make_shared<ParallelPrehandleCtx>();
 
-    for (size_t extra_id = 0; extra_id < split_key_count; extra_id++)
+    for (size_t split_id = 0; split_id < split_key_count; split_id++)
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp:535]:
```suggestion
    for (size_t split_id = 0; split_id < split_key_count; ++split_id)
```

File: dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp [hunk 2]
```diff
@@ -522,107 +523,107 @@ void executeParallelTransform(
...
         head_prehandle_result.ingest_ids.size(),
         magic_enum::enum_name(head_result.error),
-        DM::SSTScanSoftLimit::HEAD_OR_ONLY_SPLIT,
+        head_soft_limit.split_id,
         new_region->id());
 
     // Wait all threads to join. May throw.
     // If one thread throws, then all result is useless, so `async_tasks` is released directly.
-    for (size_t extra_id = 0; extra_id < split_key_count; extra_id++)
+    for (size_t split_id = 0; split_id < split_key_count; split_id++)
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp:572]:
```suggestion
    for (size_t split_id = 0; split_id < split_key_count; ++split_id)
```

File: dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp [hunk 3]
```diff
@@ -522,107 +523,107 @@ void executeParallelTransform(
...
         // Otherwise, fallback to error handling or exception handling.
         result = head_result;
         result.extra_msg = fmt::format(", from {}", DM::SSTScanSoftLimit::HEAD_OR_ONLY_SPLIT);
+        return {result, prehandle_result};
+    }
+
+    assert(head_result.error == PrehandleTransformStatus::Ok);
+    prehandle_result = std::move(head_prehandle_result);
+    // Aggregate results.
+    for (size_t split_id = 0; split_id < split_key_count; split_id++)
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp:592]:
```suggestion
    for (size_t split_id = 0; split_id < split_key_count; ++split_id)
```

**Review by Lloyd-Pottiger — APPROVED** [review]
Others LGTM

**Review by CalvinNeo — APPROVED** [review]
lgtm

**Pull request merged**